operator/pkg/networkpolicy/secretsync.go

```diff
@@ -0,0 +1,233 @@
+	if !params.NetworkPolicyConfig.EnablePolicySecretsSync {
+		return secretsync.SecretSyncRegistrationOut{}
+	}
+
+	return secretsync.SecretSyncRegistrationOut{
+		SecretSyncRegistration: &secretsync.SecretSyncRegistration{
+			RefObject:            &cilium_api_v2.CiliumClusterwideNetworkPolicy{},
+			RefObjectEnqueueFunc: EnqueueTLSSecrets(params.CtrlRuntimeManager.GetClient(), params.Logger),
+			RefObjectCheckFunc:   IsReferencedByCiliumClusterwideNetworkPolicy,
+			SecretsNamespace:     params.NetworkPolicyConfig.PolicySecretsNamespace,
+		},
+	}
+}
+
+// EnqueueTLSSecrets returns a map function that, given a CiliumNetworkPolicy or CilumClusterwideNetworkPolicy,
+// will return a slice of requests for any Secrets referenced in that CiliumNetworkPolicy.
+//
+// This includes both TLS secrets (Origination or Termination), plus Secrets used for storing header values.
+func EnqueueTLSSecrets(c client.Client, logger *slog.Logger) handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, obj client.Object) []reconcile.Request {
+		scopedLog := logger.With(logfields.Controller, "secrets", logfields.Resource, obj.GetName())
+
+		objName := types.NamespacedName{
+			Namespace: obj.GetNamespace(),
+			Name:      obj.GetName(),
+		}
+		scopedLog = scopedLog.With("networkPolicyObject", objName)
+		var specs []*api.Rule
+
+		switch o := obj.(type) {
+		case *cilium_api_v2.CiliumNetworkPolicy:
+			if o.Spec != nil {
+				specs = append(specs, o.Spec)
+			}
+			if len(o.Specs) > 0 {
+				specs = append(specs, o.Specs...)
+			}
+			scopedLog = scopedLog.With("networkPolicyKind", "CiliumNetworkPolicy")
+		case *cilium_api_v2.CiliumClusterwideNetworkPolicy:
+			if o.Spec != nil {
+				specs = append(specs, o.Spec)
+			}
+			if len(o.Specs) > 0 {
+				specs = append(specs, o.Specs...)
+			}
+			scopedLog = scopedLog.With("networkPolicyKind", "CiliumClusterwideNetworkPolicy")
+		}
+
+		var reqs []reconcile.Request
+		for _, rule := range specs {
+			for _, egress := range rule.Egress {
+				reqs = append(reqs, helpers.GetReferencedTLSSecretsFromPortRules(egress.ToPorts, scopedLog)...)
+				reqs = append(reqs, helpers.GetReferencedSecretsFromHeaderRules(egress.ToPorts, scopedLog)...)
+			}
+			for _, ingress := range rule.Ingress {
+				reqs = append(reqs, helpers.GetReferencedTLSSecretsFromPortRules(ingress.ToPorts, scopedLog)...)
+				reqs = append(reqs, helpers.GetReferencedSecretsFromHeaderRules(ingress.ToPorts, scopedLog)...)
+			}
+		}
+		return reqs
+	})
+}
+
+func IsReferencedByCiliumNetworkPolicy(ctx context.Context, c client.Client, logger *slog.Logger, obj *corev1.Secret) bool {
+	scopedLog := logger.With(logfields.Controller, "netpol-cnp-secretsync", logfields.Resource, obj.GetName())
+
+	secretName := types.NamespacedName{
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	}
+
+	cnpList := &cilium_api_v2.CiliumNetworkPolicyList{}
```

Comment:
Will this actually list against the API, or the local informer?

Reply:
My assumption was this will list using the controller-runtime pull-through cache, but I guess I could be wrong.

I'm also reluctant to make a cache unless we see problems because then we add cache management problems to the usual eventual consistency ones. We use this same approach in Gateway API handling, and I've heard from folks with thousands of HTTPRoutes that haven't noticed excessive delay time in the recalculations.

Reply:
> My assumption was this will list using the controller-runtime pull-through cache, but I guess I could be wrong.

Yes, we go with the default i.e. read from cache, write via API.